File: metropolis/platform/telemetry/python/telemetry.py
"""Metro Design telemetry client (Python).

Usage:
    from platform.telemetry.python.telemetry import Telemetry

    tel = Telemetry(dsn="...", version="1.0.0")
    tel.info("plugin loaded")
    tel.counter("effect.apply", 1, {"effect": "film_look_01"})
    tel.exception(RuntimeError("test"))
"""

from __future__ import annotations

import atexit
import logging
import time
from dataclasses import dataclass, field
from threading import Lock
from typing import Dict, Optional


class _MetricsRegistry:
    """Minimal in-memory metric registry (Prometheus exposition stub)."""

    def __init__(self) -> None:
        self._lock = Lock()
        self._counters: Dict[str, float] = {}
        self._histograms: Dict[str, list[float]] = {}

    def inc(self, name: str, value: float = 1.0) -> None:
        with self._lock:
            self._counters[name] = self._counters.get(name, 0.0) + value

    def observe(self, name: str, value: float) -> None:
        with self._lock:
            self._histograms.setdefault(name, []).append(value)

    def snapshot(self) -> dict:
        with self._lock:
            return {
                "counters": dict(self._counters),
                "histograms": {k: list(v) for k, v in self._histograms.items()},
            }


class Telemetry:
    """Telemetry client for structured logging, metrics, and error reporting."""

    def __init__(
        self,
        dsn: str = "",
        version: str = "",
        enable_sentry: bool = False,
        enable_prometheus: bool = False,
    ) -> None:
        self._version = version
        self._registry = _MetricsRegistry()
        self._logger = logging.getLogger("metropolis.telemetry")
        self._logger.setLevel(logging.INFO)

        handler = logging.StreamHandler()
        handler.setFormatter(
            logging.Formatter(
                "[%(levelname)s] %(asctime)s %(message)s",
                datefmt="%Y-%m-%dT%H:%M:%S",
            )
        )
        self._logger.addHandler(handler)
        self._logger.info("Telemetry initialized (version=%s)", version)

        self._sentry_enabled = enable_sentry
        self._prometheus_enabled = enable_prometheus

        if enable_sentry:
            self._init_sentry(dsn)

        atexit.register(self.shutdown)

    def _init_sentry(self, dsn: str) -> None:
        try:
            import sentry_sdk  # type: ignore[import-untyped]

            sentry_sdk.init(dsn=dsn, release=self._version)
            self._logger.info("Sentry integration enabled")
        except ImportError:
            self._logger.warning("sentry-sdk not installed; sentry disabled")

    def debug(self, msg: str, **extra: object) -> None:
        self._logger.debug(msg, extra=extra)

    def info(self, msg: str, **extra: object) -> None:
        self._logger.info(msg, extra=extra)

    def warn(self, msg: str, **extra: object) -> None:
        self._logger.warning(msg, extra=extra)

    def error(self, msg: str, **extra: object) -> None:
        self._logger.error(msg, extra=extra)

    def counter(
        self,
        name: str,
        value: float = 1.0,
        labels: Optional[Dict[str, str]] = None,
    ) -> None:
        self._registry.inc(name, value)

    def histogram(self, name: str, value: float) -> None:
        self._registry.observe(name, value)

    def exception(self, exc: BaseException) -> None:
        self._logger.exception(str(exc))

    def metrics_snapshot(self) -> dict:
        return self._registry.snapshot()

    def shutdown(self) -> None:
        self._logger.info("Telemetry shutting down")
        if self._sentry_enabled:
            try:
                import sentry_sdk

                sentry_sdk.flush()
            except ImportError:
                pass

File: metropolis/platform/telemetry/src/telemetry.cpp
#include "telemetry.h"
#include <cstdio>
#include <ctime>
#include <mutex>

namespace met {

struct TelemetryClient::Impl {
  std::string dsn;
  std::string version;
  bool initialized = false;
  std::mutex mtx;

  // In-memory counters for Prometheus-style exposition
  std::unordered_map<std::string, MetricSample> counters;
  std::unordered_map<std::string, MetricSample> histograms;

  void FlushLog(LogLevel level, std::string_view msg) {
    const char *level_str = "DEBUG";
    switch (level) {
      case LogLevel::kDebug:
        level_str = "DEBUG";
        break;
      case LogLevel::kInfo:
        level_str = "INFO";
        break;
      case LogLevel::kWarning:
        level_str = "WARN";
        break;
      case LogLevel::kError:
        level_str = "ERROR";
        break;
      case LogLevel::kFatal:
        level_str = "FATAL";
        break;
    }
    std::time_t now = std::time(nullptr);
    std::fprintf(stderr, "[%s] %s: %.*s\n", level_str, std::ctime(&now), static_cast<int>(msg.size()), msg.data());
  }
};

TelemetryClient::TelemetryClient() : impl_(std::make_unique<Impl>()) {}
TelemetryClient::~TelemetryClient() { Shutdown(); }

TelemetryClient &TelemetryClient::Instance() {
  static TelemetryClient instance;
  return instance;
}

void TelemetryClient::Init(std::string_view dsn, std::string_view version) {
  std::lock_guard<std::mutex> lock(impl_->mtx);
  impl_->dsn = dsn;
  impl_->version = version;
  impl_->initialized = true;
  impl_->FlushLog(LogLevel::kInfo, "Telemetry initialized");
}

void TelemetryClient::Shutdown() {
  std::lock_guard<std::mutex> lock(impl_->mtx);
  if (impl_->initialized) {
    impl_->FlushLog(LogLevel::kInfo, "Telemetry shutting down");
    impl_->initialized = false;
  }
}

void TelemetryClient::Log(LogLevel level, std::string_view message) {
  std::lock_guard<std::mutex> lock(impl_->mtx);
  impl_->FlushLog(level, message);
}

void TelemetryClient::LogInfo(std::string_view message) { Log(LogLevel::kInfo, message); }
void TelemetryClient::LogWarn(std::string_view message) { Log(LogLevel::kWarning, message); }
void TelemetryClient::LogError(std::string_view message) { Log(LogLevel::kError, message); }

void TelemetryClient::IncCounter(std::string_view name, double value,
                                 std::unordered_map<std::string, std::string> labels) {
  std::lock_guard<std::mutex> lock(impl_->mtx);
  std::string key(name);
  auto it = impl_->counters.find(key);
  if (it != impl_->counters.end()) {
    it->second.value += value;
  } else {
    impl_->counters[key] = MetricSample{std::string(name), value, std::move(labels)};
  }
}

void TelemetryClient::ObserveHistogram(std::string_view name, double value,
                                       std::unordered_map<std::string, std::string> labels) {
  std::lock_guard<std::mutex> lock(impl_->mtx);
  std::string key(name);
  impl_->histograms[key] = MetricSample{std::string(name), value, std::move(labels)};
}

void TelemetryClient::CaptureException(std::string_view message) {
  Log(LogLevel::kError, message);
}

void TelemetryClient::SetUser(std::string_view user_id, std::string_view email) {
  LogInfo(std::string("User: ") + std::string(user_id));
}

}  // namespace met
